File: frontend/src/types/active-learning.ts
export type TrainingRunStatus = "PENDING" | "RUNNING" | "SUCCESS" | "FAILED";
export type PredictionRunStatus = "PENDING" | "RUNNING" | "SUCCESS" | "FAILED";
export type CoresetRunStatus = "PENDING" | "RUNNING" | "SUCCESS" | "FAILED";
export type ConfidenceStrategy =
  | "mean_max_probability"
  | "min_max_probability"
  | "entropy"
  | "margin";
export type ActiveLearningStage = "junior" | "senior";
export type CoresetStrategy =
  | "random"
  | "kmeans"
  | "facility_location"
  | "graph_cut"
  | "moe";
export type CoresetDataScope = "all" | "pending";
export type CoresetUpdateMode = "full" | "incremental";

export type TrainingRunCreateRequest = {
  triggered_by_user_id?: number | null;
  target_stage: ActiveLearningStage;
  epochs: number;
  batch_size: number;
  learning_rate: number;
  val_size: number;
  patience: number;
  max_length: number;
  random_seed: number;
  include_gold_labels: boolean;
  min_train_samples: number;
  device: "auto" | "cpu" | "cuda";
};

export type TrainingEpoch = {
  id: number;
  training_run_id: number;
  epoch_no: number;
  train_loss?: number | null;
  val_loss?: number | null;
  level_accuracy?: number | null;
  macro_f1?: number | null;
  detection_rate?: number | null;
  created_at: string;
};

export type TrainingRun = {
  id: number;
  run_no: string;
  run_display_name: string;
  status: TrainingRunStatus;
  triggered_by_user_id?: number | null;
  base_model_path: string;
  target_stage: string;
  train_sample_count: number;
  val_sample_count: number;
  dataset_sample_count: number;
  model_type?: string | null;
  base_model_name?: string | null;
  parameter_summary: string;
  trend_group_key: string;
  related_model_version_id?: number | null;
  related_model_version_code?: string | null;
  related_model_display_name?: string | null;
  params_json?: Record<string, unknown> | null;
  metrics_json?: Record<string, number> | null;
  error_message?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  epochs: TrainingEpoch[];
};

export type TrainingRunLog = {
  run_id: number;
  log_text: string;
  stderr_text: string;
  is_truncated: boolean;
};

export type ModelVersion = {
  id: number;
  version_code: string;
  version_display_name: string;
  model_type?: string | null;
  base_model_name?: string | null;
  artifact_path?: string | null;
  training_run_id: number;
  source_run_no?: string | null;
  source_run_display_name?: string | null;
  checkpoint_path: string;
  is_active: boolean;
  level_accuracy?: number | null;
  macro_f1?: number | null;
  detection_rate?: number | null;
  val_loss?: number | null;
  train_sample_count: number;
  val_sample_count: number;
  dataset_sample_count: number;
  parameter_summary: string;
  trend_group_key: string;
  params_json?: Record<string, unknown> | null;
  created_at: string;
};

export type TrendPoint = {
  model_version_id: number;
  training_run_id: number;
  label: string;
  sample_label: string;
  sample_count: number;
  train_sample_count: number;
  val_sample_count: number;
  level_accuracy?: number | null;
  macro_f1?: number | null;
  detection_rate?: number | null;
  created_at: string;
};

export type TrendGroup = {
  key: string;
  label: string;
  parameter_summary: string;
  target_stage: string;
  model_type?: string | null;
  base_model_name?: string | null;
  point_count: number;
  points: TrendPoint[];
};

export type PredictionRunCreateRequest = {
  triggered_by_user_id?: number | null;
  model_version_id?: number | null;
  target_stage: ActiveLearningStage;
  select_count: number;
  confidence_strategy: ConfidenceStrategy;
  batch_size: number;
  auto_move_to_waiting: boolean;
};

export type PredictionItem = {
  id: number;
  prediction_run_id: number;
  question_id: number;
  predicted_levels_json: number[];
  confidence_score: number;
  uncertainty_score: number;
  rank_no: number;
  is_selected: boolean;
  created_at: string;
};

export type PredictionRun = {
  id: number;
  run_no: string;
  model_version_id: number;
  status: PredictionRunStatus;
  triggered_by_user_id?: number | null;
  confidence_strategy: ConfidenceStrategy;
  candidate_count: number;
  selected_count: number;
  moved_count: number;
  recommendation_batch_id?: number | null;
  params_json?: Record<string, unknown> | null;
  metrics_json?: Record<string, number | null> | null;
  error_message?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  items: PredictionItem[];
};

export type CoresetRunCreateRequest = {
  triggered_by_user_id?: number | null;
  strategy: CoresetStrategy;
  count: number;
  data_scope: CoresetDataScope;
  update_mode: CoresetUpdateMode;
};

export type CoresetIncrementalSummary = {
  can_run_incremental: boolean;
  data_scope: CoresetDataScope;
  baseline_run_id?: number | null;
  baseline_run_no?: string | null;
  baseline_batch_no?: string | null;
  baseline_strategy?: CoresetStrategy | null;
  baseline_finished_at?: string | null;
  baseline_selected_count: number;
  current_pool_count: number;
  new_unlabeled_count: number;
  incremental_candidate_count: number;
  anchor_count: number;
  snapshot_created_before?: string | null;
};

export type CoresetRun = {
  id: number;
  run_no: string;
  status: CoresetRunStatus;
  triggered_by_user_id?: number | null;
  strategy: CoresetStrategy;
  data_scope: CoresetDataScope;
  update_mode: CoresetUpdateMode;
  requested_count: number;
  candidate_count: number;
  selected_count: number;
  moved_count: number;
  recommendation_batch_id?: number | null;
  batch_no?: string | null;
  recommendation_batch_no?: string | null;
  params_json?: Record<string, unknown> | null;
  metrics_json?: Record<string, unknown> | null;
  error_message?: string | null;
  baseline_run_id?: number | null;
  baseline_run_no?: string | null;
  baseline_batch_no?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  question_ids: number[];
  moved_question_ids: number[];
};

export type ActiveLearningOverviewResponse = {
  active_model?: ModelVersion | null;
  model_versions: ModelVersion[];
  training_runs: TrainingRun[];
  prediction_runs: PredictionRun[];
  coreset_runs: CoresetRun[];
  coreset_incremental?: CoresetIncrementalSummary | null;
  coreset_incremental_by_strategy?: Record<string, CoresetIncrementalSummary>;
  trend_groups: TrendGroup[];
  completed_sample_count: number;
  pending_candidate_count: number;
};

File: backend/app/schemas/active_learning.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

TrainingRunStatus = Literal["PENDING", "RUNNING", "SUCCESS", "FAILED"]
PredictionRunStatus = Literal["PENDING", "RUNNING", "SUCCESS", "FAILED"]
CoresetRunStatus = Literal["PENDING", "RUNNING", "SUCCESS", "FAILED"]
ActiveLearningStage = Literal["junior", "senior"]
ConfidenceStrategy = Literal["mean_max_probability", "min_max_probability", "entropy", "margin"]
TrainingDevice = Literal["auto", "cpu", "cuda"]
SelectionStrategy = Literal["random", "kmeans", "facility_location", "graph_cut", "moe"]
SelectionDataScope = Literal["all", "pending"]
CoresetUpdateMode = Literal["full", "incremental"]


class TrainingRunCreateRequest(BaseModel):
    triggered_by_user_id: int | None = None
    target_stage: ActiveLearningStage = "junior"
    epochs: int = Field(default=20, ge=1, le=50)
    batch_size: int = Field(default=16, ge=1, le=128)
    learning_rate: float = Field(default=5e-5, gt=0, le=1)
    val_size: float = Field(default=0.2, ge=0.05, le=0.5)
    patience: int = Field(default=5, ge=1, le=50)
    max_length: int = Field(default=256, ge=32, le=512)
    random_seed: int = Field(default=42, ge=0, le=1_000_000)
    include_gold_labels: bool = False
    min_train_samples: int = Field(default=5, ge=1, le=10000)
    device: TrainingDevice = "auto"
    max_coreset_round: int | None = Field(default=None, ge=1)


class TrainingEpochOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    training_run_id: int
    epoch_no: int
    train_loss: float | None = None
    val_loss: float | None = None
    level_accuracy: float | None = None
    macro_f1: float | None = None
    detection_rate: float | None = None
    created_at: datetime


class TrainingRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_no: str
    status: TrainingRunStatus
    triggered_by_user_id: int | None = None
    base_model_path: str
    target_stage: str
    train_sample_count: int
    val_sample_count: int
    dataset_sample_count: int
    run_display_name: str
    model_type: str | None = None
    base_model_name: str | None = None
    parameter_summary: str
    trend_group_key: str
    related_model_version_id: int | None = None
    related_model_version_code: str | None = None
    related_model_display_name: str | None = None
    params_json: dict | None = None
    metrics_json: dict | None = None
    error_message: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    epochs: list[TrainingEpochOut] = Field(default_factory=list)


class TrainingRunLogOut(BaseModel):
    run_id: int
    log_text: str
    stderr_text: str = ""
    is_truncated: bool = False


class ModelVersionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    version_code: str
    version_display_name: str
    model_type: str | None = None
    base_model_name: str | None = None
    artifact_path: str | None = None
    training_run_id: int
    source_run_no: str | None = None
    source_run_display_name: str | None = None
    checkpoint_path: str
    is_active: bool
    level_accuracy: float | None = None
    macro_f1: float | None = None
    detection_rate: float | None = None
    val_loss: float | None = None
    train_sample_count: int
    val_sample_count: int
    dataset_sample_count: int
    parameter_summary: str
    trend_group_key: str
    params_json: dict | None = None
    created_at: datetime


class TrendPointOut(BaseModel):
    model_version_id: int
    training_run_id: int
    label: str
    sample_label: str
    sample_count: int
    train_sample_count: int
    val_sample_count: int
    level_accuracy: float | None = None
    macro_f1: float | None = None
    detection_rate: float | None = None
    created_at: datetime


class TrendGroupOut(BaseModel):
    key: str
    label: str
    parameter_summary: str
    target_stage: str
    model_type: str | None = None
    base_model_name: str | None = None
    point_count: int
    points: list[TrendPointOut] = Field(default_factory=list)


class PredictionRunCreateRequest(BaseModel):
    triggered_by_user_id: int | None = None
    model_version_id: int | None = None
    target_stage: ActiveLearningStage = "junior"
    select_count: int = Field(default=100, ge=1, le=5000)
    confidence_strategy: ConfidenceStrategy = "mean_max_probability"
    batch_size: int = Field(default=32, ge=1, le=256)
    auto_move_to_waiting: bool = True


class PredictionItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    prediction_run_id: int
    question_id: int
    predicted_levels_json: list[int]
    confidence_score: float
    uncertainty_score: float
    rank_no: int
    is_selected: bool
    created_at: datetime


class PredictionRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_no: str
    model_version_id: int
    status: PredictionRunStatus
    triggered_by_user_id: int | None = None
    confidence_strategy: str
    candidate_count: int
    selected_count: int
    moved_count: int
    recommendation_batch_id: int | None = None
    params_json: dict | None = None
    metrics_json: dict | None = None
    error_message: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    items: list[PredictionItemOut] = Field(default_factory=list)


class CoresetRunCreateRequest(BaseModel):
    triggered_by_user_id: int | None = None
    strategy: SelectionStrategy = "kmeans"
    count: int = Field(default=400, ge=1, le=5000)
    data_scope: SelectionDataScope = "pending"
    update_mode: CoresetUpdateMode = "full"


class CoresetIncrementalSummaryOut(BaseModel):
    can_run_incremental: bool
    data_scope: SelectionDataScope = "pending"
    baseline_run_id: int | None = None
    baseline_run_no: str | None = None
    baseline_batch_no: str | None = None
    baseline_strategy: SelectionStrategy | None = None
    baseline_finished_at: datetime | None = None
    baseline_selected_count: int = 0
    current_pool_count: int = 0
    new_unlabeled_count: int = 0
    incremental_candidate_count: int = 0
    anchor_count: int = 0
    snapshot_created_before: datetime | None = None


class CoresetRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_no: str
    status: CoresetRunStatus
    triggered_by_user_id: int | None = None
    strategy: SelectionStrategy
    data_scope: SelectionDataScope
    update_mode: CoresetUpdateMode = "full"
    requested_count: int
    candidate_count: int
    selected_count: int
    moved_count: int
    recommendation_batch_id: int | None = None
    batch_no: str | None = None
    recommendation_batch_no: str | None = None
    params_json: dict | None = None
    metrics_json: dict | None = None
    error_message: str | None = None
    baseline_run_id: int | None = None
    baseline_run_no: str | None = None
    baseline_batch_no: str | None = None
    active_learning_round: int | None = None
    round_completed_count: int = 0
    round_unfinished_count: int = 0
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    question_ids: list[int] = Field(default_factory=list)
    moved_question_ids: list[int] = Field(default_factory=list)


class ActiveLearningOverviewResponse(BaseModel):
    active_model: ModelVersionOut | None = None
    model_versions: list[ModelVersionOut]
    training_runs: list[TrainingRunOut]
    prediction_runs: list[PredictionRunOut]
    coreset_runs: list[CoresetRunOut] = Field(default_factory=list)
    coreset_incremental: CoresetIncrementalSummaryOut | None = None
    coreset_incremental_by_strategy: dict[str, CoresetIncrementalSummaryOut] = Field(default_factory=dict)
    trend_groups: list[TrendGroupOut] = Field(default_factory=list)
    completed_sample_count: int
    pending_candidate_count: int
